safe_app/src/tests/mod.rs

```diff
@@ -32,14 +31,137 @@ use safe_core::{Client, CoreError};
 use safe_nd::{
     ADataAddress, ADataOwner, AppPermissions, AppendOnlyData, Coins, Error as SndError,
     PubImmutableData, PubSeqAppendOnlyData, PubUnseqAppendOnlyData, UnpubUnseqAppendOnlyData,
-    XorName,
+    XorName, GET_BALANCE, PERFORM_MUTATIONS, TRANSFER_COINS,
 };
 #[cfg(feature = "mock-network")]
 use safe_nd::{RequestType, Response};
 use std::collections::HashMap;
 use std::rc::Rc;
 use unwrap::unwrap;
 
+// App Tokens test:
+// 1. Authenticate an app A.
+// 2. Send request as app A assert that tokens are valid.
+// 3. Re-authenticate app A again with new permissions so that token gets updated.
+// 4. Assert if the tokens are updated by sending the older token - should fail.
+// 5. Assert again with the new token - should pass.
+#[test]
+pub fn app_tokens_test() {
+    let auth = authenticator::create_account_and_login();
+    let app_info = authenticator::rand_app();
+
+    let app_permissions1 = AppPermissions {
+        transfer_coins: true,
+        perform_mutations: true,
+        get_balance: true,
+    };
+
+    // Authenticate App A
+    let auth_response = unwrap!(authenticator::register_app(
+        &auth,
+        &AuthReq {
+            app: app_info.clone(),
+            app_container: false,
+            app_permissions: app_permissions1,
+            containers: HashMap::new(),
+        },
+    ));
+
+    let old_token = auth_response.clone().token;
+    // Verify the token
+    unwrap!(auth_run(&auth, move |client| {
+        // Verify if the signature is signed by the client and is verifiable by client's public ID
+        assert!(old_token
+            .is_valid_for_public_key(&client.public_id().public_key())
+            .is_ok());
+
+        // Verify the contents in the Caveat
+        assert!(old_token
+            .verify_caveat(GET_BALANCE, |content| content == "true")
+            .is_ok());
+
+        assert!(old_token
+            .verify_caveat(PERFORM_MUTATIONS, |content| content == "true")
+            .is_ok());
+
+        assert!(old_token
+            .verify_caveat(TRANSFER_COINS, |content| content == "true")
+            .is_ok());
+        Ok(())
+    }));
+
+    let app = unwrap!(App::registered(app_info.id.clone(), auth_response, || ()));
+
+    // Send requests to check app permissions before updating token - should pass
+    unwrap!(run(&app, move |client, _| {
+        client
+            .get_balance(None)
+            .map_err(AppError::from)
+            .then(move |res| {
+                match res {
+                    Ok(_) => (),
+                    Err(e) => panic!("Unexpected Error: {:?}", e),
+                }
+                Ok(())
+            })
+    }));
+
+    // Re-authenticate the app with new permissions
+    let app_permissions2 = AppPermissions {
+        transfer_coins: false,
```

Comment:
Only tweak I'd make here is that, in this second instance, we should have _different_ permissions, but still check the same one. That way we know that the access denied is due to token hash, and _not_ the permission itself.